pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	gardenhelper "github.com/gardener/gardener/pkg/apis/garden/v1beta1/helper"
+	gardeninformers "github.com/gardener/gardener/pkg/client/garden/informers/externalversions"
+	gardenlisters "github.com/gardener/gardener/pkg/client/garden/listers/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/scheduler/apis/config"
+	schedulerutils "github.com/gardener/gardener/pkg/scheduler/utils"
+	"github.com/gardener/gardener/pkg/utils"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+func (c *SchedulerController) shootAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.shootQueue.Add(key)
+}
+
+func (c *SchedulerController) shootUpdate(oldObj, newObj interface{}) {
+	var (
+		oldShoot      = oldObj.(*gardenv1beta1.Shoot)
+		newShoot      = newObj.(*gardenv1beta1.Shoot)
+		specChanged   = !apiequality.Semantic.DeepEqual(oldShoot.Spec, newShoot.Spec)
+		statusChanged = !apiequality.Semantic.DeepEqual(oldShoot.Status, newShoot.Status)
+	)
+
+	if !specChanged && statusChanged {
```

Comment:
Why is that needed? can't we directly ignore if the `.spec.cloud.seed` is non-empty?